src/provider.rs

```diff
@@ -34,30 +40,49 @@ use crate::blobs::{Blob, Collection};
 use crate::protocol::{
     read_lp, write_lp, AuthToken, Closed, Handshake, Request, Res, Response, VERSION,
 };
+use crate::rpc_protocol::{
+    ListRequest, ListResponse, ProvideRequest, ProvideResponse, ProviderRequest, ProviderService,
+    VersionRequest, VersionResponse, WatchRequest, WatchResponse,
+};
 use crate::tls::{self, Keypair, PeerId};
 use crate::util::{self, Hash};
 
 const MAX_CONNECTIONS: u32 = 1024;
 const MAX_STREAMS: u64 = 10;
+const HEALTH_POLL_WAIT: Duration = Duration::from_secs(1);
 
 /// Database containing content-addressed data (blobs or collections).
-#[derive(Debug, Clone)]
-pub struct Database(Arc<HashMap<Hash, BlobOrCollection>>);
+#[derive(Debug, Clone, Default)]
+pub struct Database(Arc<Mutex<HashMap<Hash, BlobOrCollection>>>);
 
 impl Database {
-    fn get(&self, key: &Hash) -> Option<&BlobOrCollection> {
-        self.0.get(key)
+    fn get(&self, key: &Hash) -> Option<BlobOrCollection> {
+        self.0.lock().unwrap().get(key).cloned()
+    }
+
+    fn union_with(&self, db: HashMap<Hash, BlobOrCollection>) {
+        let mut inner = self.0.lock().unwrap();
+        for (k, v) in db {
+            inner.entry(k).or_insert(v);
+        }
     }
 
     /// Iterate over all blobs in the database.
-    pub fn blobs(&self) -> impl Iterator<Item = (&Hash, &PathBuf, u64)> + '_ {
-        self.0
+    pub fn blobs(&self) -> impl Iterator<Item = (Hash, PathBuf, u64)> + 'static {
+        let items = self
+            .0
+            .lock()
+            .unwrap()
             .iter()
             .filter_map(|(k, v)| match v {
                 BlobOrCollection::Blob(data) => Some((k, data)),
                 BlobOrCollection::Collection(_) => None,
             })
-            .map(|(k, data)| (k, &data.path, data.size))
+            .map(|(k, data)| (*k, data.path.clone(), data.size))
+            .collect::<Vec<_>>();
```

Comment:
This`.collect()` is pretty unfortunate, but I guess the lock forces our hand.

Reply:
If you don't want to .collect you would have to use an immutable collection for the db that has zero cost snapshots. Like https://crates.io/crates/im . But I think we can keep it like this.

In the long term this will have to be a database anyway, maybe redb...